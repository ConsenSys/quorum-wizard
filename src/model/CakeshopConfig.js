import {
  isDocker,
  isTessera,
} from './NetworkConfig'

// eslint-disable-next-line import/prefer-default-export
export function generateCakeshopConfig(config) {
  const nodes = []

  config.nodes.forEach((node, i) => {
    const nodeData = {
      name: `node${i + 1}`,
      rpcUrl: isDocker(config.network.deployment)
        // docker-compose creates dns entries for the container's ip using container name (nodeX)
        ? `http://node${i + 1}:${config.containerPorts.quorum.rpcPort}`
        : `http://localhost:${node.quorum.rpcPort}`,
      transactionManagerUrl: createTransactionManagerUrl(config, node, i),
    }
    nodes.push(nodeData)
  })
  return nodes
}

function createTransactionManagerUrl(config, node, i) {
  if (isTessera(config.network.transactionManager)) {
    return isDocker(config.network.deployment)
      // docker-compose creates dns entries for the container's ip using container name (txmanagerX)
      ? `http://txmanager${i + 1}:${config.containerPorts.tm.thirdPartyPort}`
      : `http://localhost:${node.tm.thirdPartyPort}`
  }
  return ''
}
