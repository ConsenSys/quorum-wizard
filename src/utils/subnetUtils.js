const ipaddress = require('ip-address')
const { BigInteger } = require('jsbn')

// eslint-disable-next-line import/prefer-default-export
export function cidrsubnet(cidr, newBits, netNum) {
  const ipv4 = new ipaddress.Address4(cidr)

  const newMask = ipv4.subnetMask + newBits

  if (newMask > 32) {
    throw new Error(`Requested ${newBits}  new bits, but only ${32 - ipv4.subnetMask} are available.`)
  }

  const maxNetNum = 1 << newBits // eslint-disable-line no-bitwise

  if (netNum > maxNetNum) {
    throw new Error(`prefix extension of ${newBits} does not accommodate a subnet numbered ${netNum}`)
  }

  const addrBig = ipv4.bigInteger()
  const netNumBig = new BigInteger(netNum.toString())
  const addToIp = netNumBig << (32 - newMask) // eslint-disable-line no-bitwise
  const newAddrBig = addrBig.add(new BigInteger(addToIp.toString()))
  const newAddr = ipaddress.Address4.fromBigInteger(newAddrBig).address
  return `${newAddr}/${newMask}`
}

export function cidrhost(cidr, hostNum) {
  const ipv4 = new ipaddress.Address4(cidr)
  const hostNumBig = new BigInteger(hostNum.toString())

  const parentLen = ipv4.subnetMask
  const hostLen = 32 - parentLen

  // calculate max host num
  let maxHostNum = new BigInteger('1')
  maxHostNum <<= new BigInteger(hostLen.toString()) // eslint-disable-line no-bitwise
  maxHostNum -= new BigInteger('1')

  if (hostNum > maxHostNum) {
    throw new Error(`prefix of ${parentLen} does not accommodate a host numbered ${hostNum}`)
  }
  const startingAddrBig = ipv4.startAddress().bigInteger()
  const newAddrBig = startingAddrBig.add(hostNumBig)

  return ipaddress.Address4.fromBigInteger(newAddrBig).address
}

export function getRandomInt(min, max) {
  // The maximum is exclusive and the minimum is inclusive
  const minCeil = Math.ceil(min)
  const maxFloor = Math.floor(max)
  return Math.floor(Math.random() * (maxFloor - minCeil)) + minCeil
}

export function getDockerSubnet() {
  // min=1, max=254
  const randomNetNum = getRandomInt(1, 255)
  return cidrsubnet('172.16.0.0/16', 8, randomNetNum)
}

export function buildDockerIp(ip, end) {
  const ipArray = ip.split('.').slice(0, 3)
  ipArray.push(end)
  const dockerIp = ipArray.join('.')
  return dockerIp
}
