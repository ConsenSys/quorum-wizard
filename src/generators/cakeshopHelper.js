import { join } from 'path'
import {
  copyFile,
  createFolder,
  cwd,
  writeJsonFile,
} from '../utils/fileUtils'
import { generateCakeshopConfig } from '../model/CakeshopConfig'

export function buildCakeshopDir(config, qdata) {
  const cakeshopDir = join(qdata, 'cakeshop', 'local')
  createFolder(cakeshopDir, true)
  writeJsonFile(cakeshopDir, 'cakeshop.json', generateCakeshopConfig(config))
  copyFile(join(cwd(), 'lib/cakeshop_application.properties.template'), join(cakeshopDir, 'application.properties'))
}

export function generateCakeshopScript(qdata) {
  const jvmParams = "-Dcakeshop.config.dir=qdata/cakeshop -Dlogging.path=qdata/logs/cakeshop"
  const CMD = `java ${jvmParams} -jar $BIN_CAKESHOP > /dev/null 2>&1 &`
  return CMD
}

export function waitForCakeshopCommand(config) {
  return `
  DOWN=true
  k=10
  while \${DOWN}; do
    sleep 1
    echo "Waiting until Cakeshop is running..."
    DOWN=false
    set +e
    result=$(curl -s http://localhost:8999/actuator/health)
    set -e
    if [ ! "\${result}" == "{\\"status\\":\\"UP\\"}" ]; then
      echo "Cakeshop is not yet listening on http"
      DOWN=true
    fi

    k=$((k-1))
    if [ \${k} -le 0 ]; then
      echo "Cakeshop is taking a long time to start. Look at logs"
    fi

    sleep 5
  done

  echo "Cakeshop started at http://localhost:8999"
  exit 0
  `
}
