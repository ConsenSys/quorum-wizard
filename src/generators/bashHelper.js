import {
  getFullNetworkPath,
} from './networkCreator'
import {
  copyFile,
  libRootDir,
  writeFile,
} from '../utils/fileUtils'
import { executeSync } from '../utils/execUtils'
import {
  buildCakeshopDir,
  generateCakeshopScript,
} from './cakeshopHelper'
import {
  isQuorum260Plus,
  pathToCakeshop,
  pathToQuorumBinary,
  pathToTesseraJar,
} from './binaryHelper'
import {
  isRaft,
  isTessera,
  isCakeshop,
} from '../model/NetworkConfig'
import { info } from '../utils/log'
import { formatTesseraKeysOutput } from './transactionManager'
import { joinPath } from '../utils/pathUtils'

export function buildBashScript(config) {
  const commands = createCommands(config)

  const startScript = [
    '#!/bin/bash',
    'echo "\nStarting Quorum network...\n"',
    setEnvironmentCommand(config),
    commands.tesseraStart,
    waitForTesseraNodesCommand(config),
    'echo "Starting Quorum nodes"',
    commands.gethStart,
    generateCakeshopScript(config),
    'echo "Successfully started Quorum network."',
    `echo "${formatTesseraKeysOutput(config)}"`,
  ]

  return {
    startScript: startScript.join('\n'),
    initCommands: commands.initStart,
  }
}

export function createCommands(config) {
  const networkPath = getFullNetworkPath(config)
  const initCommands = []
  const startCommands = []
  const tmStartCommands = []

  config.nodes.forEach((node, i) => {
    const nodeNumber = i + 1
    const quorumDir = joinPath('qdata', `dd${nodeNumber}`)
    const tmDir = joinPath('qdata', `c${nodeNumber}`)
    const genesisLocation = joinPath(quorumDir, 'genesis.json')
    const keyDir = joinPath(quorumDir, 'keystore')
    const passwordDestination = joinPath(keyDir, 'password.txt')
    const logs = joinPath('qdata', 'logs')
    const initCommand = `cd ${networkPath} && ${pathToQuorumBinary(config.network.quorumVersion)} --datadir ${quorumDir} init ${genesisLocation} 2>&1`
    initCommands.push(initCommand)

    const tmIpcLocation = isTessera(config.network.transactionManager)
      ? joinPath(tmDir, 'tm.ipc')
      : 'ignore'
    const startCommand = createGethStartCommand(
      config,
      node,
      passwordDestination,
      nodeNumber,
      tmIpcLocation,
    )
    startCommands.push(startCommand)

    if (isTessera(config.network.transactionManager)) {
      const tmStartCommand = createTesseraStartCommand(config, node, nodeNumber, tmDir, logs)
      tmStartCommands.push(tmStartCommand)
    }
  })

  const obj = {
    tesseraStart: tmStartCommands.join('\n'),
    gethStart: startCommands.join('\n'),
    initStart: initCommands,
  }
  return obj
}


export async function buildBash(config) {
  const bashDetails = buildBashScript(config)
  const networkPath = getFullNetworkPath(config)

  if (isCakeshop(config.network.cakeshop)) {
    buildCakeshopDir(config, joinPath(networkPath, 'qdata'))
  }

  info('Writing start script...')
  writeFile(joinPath(networkPath, 'start.sh'), bashDetails.startScript, true)
  copyFile(joinPath(libRootDir(), 'lib', 'stop.sh'), joinPath(networkPath, 'stop.sh'))

  info('Initializing quorum...')
  bashDetails.initCommands.forEach((command) => executeSync(command))
  info('Done')
}

export function createGethStartCommand(config, node, passwordDestination, nodeNumber, tmIpcPath) {
  const {
    verbosity, networkId, consensus, quorumVersion,
  } = config.network
  const {
    devP2pPort, rpcPort, raftPort, wsPort,
  } = node.quorum
  const quorum26Flags = isQuorum260Plus(quorumVersion) ? '--allow-insecure-unlock' : ''

  const args = `--nodiscover --rpc --rpccorsdomain=* --rpcvhosts=* --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum,${consensus},quorumPermission --ws --wsaddr 0.0.0.0 --wsorigins=* --emitcheckpoints --unlock 0 --password ${passwordDestination} ${quorum26Flags}`
  const consensusArgs = isRaft(consensus)
    ? `--raft --raftport ${raftPort}`
    : '--istanbul.blockperiod 5 --syncmode full --mine --minerthreads 1'

  return `PRIVATE_CONFIG=${tmIpcPath} nohup $BIN_GETH --datadir qdata/dd${nodeNumber} ${args} ${consensusArgs} --permissioned --verbosity ${verbosity} --networkid ${networkId} --rpcport ${rpcPort} --wsport ${wsPort} --port ${devP2pPort} 2>>qdata/logs/${nodeNumber}.log &`
}

export function createTesseraStartCommand(config, node, nodeNumber, tmDir, logDir) {
  // `rm -f ${tmDir}/tm.ipc`

  let DEBUG = ''
  if (config.network.remoteDebug) {
    DEBUG = '-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=500$i -Xdebug'
  }

  const MEMORY = '-Xms128M -Xmx128M'
  const CMD = `java ${DEBUG} ${MEMORY} -jar $BIN_TESSERA -configfile ${tmDir}/tessera-config-09-${nodeNumber}.json >> ${logDir}/tessera${nodeNumber}.log 2>&1 &`
  return CMD
}

function checkTesseraUpcheck(nodes) {
  return nodes.map((node, i) => `
    result${i + 1}=$(curl -s http://${node.tm.ip}:${node.tm.p2pPort}/upcheck)
    if [ ! "\${result${i + 1}}" == "I'm up!" ]; then
        echo "Node ${i + 1} is not yet listening on http"
        DOWN=true
    fi`)
}

export function setEnvironmentCommand(config) {
  const lines = []
  lines.push(`BIN_GETH=${pathToQuorumBinary(config.network.quorumVersion)}`)
  if (isTessera(config.network.transactionManager)) {
    lines.push(`BIN_TESSERA=${pathToTesseraJar(config.network.transactionManager)}`)
  }
  if (isCakeshop(config.network.cakeshop)) {
    lines.push(`BIN_CAKESHOP=${pathToCakeshop(config.network.cakeshop)}`)
  }
  lines.push('')
  return lines.join('\n')
}

export function waitForTesseraNodesCommand(config) {
  if (!isTessera(config.network.transactionManager)) {
    return ''
  }
  return `
echo "Waiting until all Tessera nodes are running..."
DOWN=true
k=10
while \${DOWN}; do
    sleep 1
    DOWN=false
    for i in \`seq 1 ${config.nodes.length}\`
    do
        if [ ! -S "qdata/c\${i}/tm.ipc" ]; then
            echo "Node \${i} is not yet listening on tm.ipc"
            DOWN=true
        fi
    done
    set +e
    #NOTE: if using https, change the scheme
    #NOTE: if using the IP whitelist, change the host to an allowed host
    ${checkTesseraUpcheck(config.nodes).join('')}

    k=$((k - 1))
    if [ \${k} -le 0 ]; then
        echo "Tessera is taking a long time to start.  Look at the Tessera logs in qdata/logs/ for help diagnosing the problem."
    fi
    echo "Waiting until all Tessera nodes are running..."

    sleep 5
done

echo "All Tessera nodes started"
`
}
