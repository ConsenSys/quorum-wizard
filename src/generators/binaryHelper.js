import cmp from 'semver-compare'
import { wizardHomeDir } from '../utils/fileUtils'
import { executeSync } from '../utils/execUtils'
import {
  isBash,
  isIstanbul,
  isTessera,
  isKubernetes,
} from '../model/NetworkConfig'
import {
  BINARIES,
  downloadIfMissing,
  LATEST_BOOTNODE,
  LATEST_ISTANBUL_TOOLS,
} from './download'
import { disableIfWrongJavaVersion } from '../questions/validators'
import { info } from '../utils/log'
import { joinPath } from '../utils/pathUtils'

// This method could be improved, but right now it tries to:
// a. Cache downloads
// b. Only download if necessary for bash deployments
export async function downloadAndCopyBinaries(config) {
  info('Downloading dependencies...')
  const {
    transactionManager, cakeshop, generateKeys, quorumVersion, consensus,
  } = config.network

  const downloads = []

  if (isIstanbul(consensus)) {
    downloads.push(downloadIfMissing('istanbul', LATEST_ISTANBUL_TOOLS))
  }

  if (generateKeys) {
    downloads.push(downloadIfMissing('bootnode', LATEST_BOOTNODE))
  }

  if (quorumVersion !== 'PATH') {
    downloads.push(downloadIfMissing('quorum', quorumVersion))
  }
  const tesseraVersion = transactionManager
  if (tesseraVersion !== 'PATH' && isTessera(tesseraVersion)) {
    downloads.push(downloadIfMissing('tessera', tesseraVersion))
  }

  if (cakeshop !== 'none') {
    downloads.push(downloadIfMissing('cakeshop', cakeshop))
  }

  await Promise.all(downloads)
}

export function getGethOnPath() {
  const pathChoices = []
  try {
    const gethOnPath = executeSync('which geth').toString().trim()
    if (gethOnPath) {
      const version = getPathGethVersion()
      if (version !== null) {
        pathChoices.push({
          name: `Quorum ${version} on path (${gethOnPath})`,
          value: 'PATH',
        })
      }
    }
  } catch (e) {
    // either no geth or the version call errored, don't include it in choices
  }
  return pathChoices
}

export function getPathGethVersion() {
  const gethVersionOutput = executeSync('geth version').toString()
  const versionMatch = gethVersionOutput.match(/Quorum Version: (\S+)/)
  if (versionMatch !== null) {
    return versionMatch[1]
  }
  return null
}

export function isQuorum260Plus(quorumVersion) {
  const version = quorumVersion === 'PATH' ? getPathGethVersion() : quorumVersion
  return cmp(version, '2.6.0') >= 0
}

export function isQuorum2010Plus(quorumVersion) {
  const version = quorumVersion === 'PATH' ? getPathGethVersion() : quorumVersion
  return cmp(version, '20.10.0') >= 0
}

export function isLegacyTessera(tesseraVersion) {
  if (tesseraVersion === 'PATH') {
    // no easy way to get version from tessera jar, assume it is not legacy
    return false
  }
  if (tesseraVersion === 'none') {
    // treat no tessera as legacy, since this determines if enhanced privacy is enabled
    return true
  }
  return cmp(tesseraVersion, '1.0.0') < 0
}

export function getDownloadableGethChoices(deployment) {
  let choices = getDownloadableChoices(BINARIES.quorum)
  if (isBash(deployment)) {
    choices = choices.concat(getGethOnPath())
  }
  return choices
}

export function getDownloadableTesseraChoices(deployment) {
  let choices = getDownloadableChoices(BINARIES.tessera)
  if (isBash(deployment)) {
    choices = choices.concat(getTesseraOnPath())
  } else {
    // allow all options in docker compose mode since local jdk version doesn't matter
    choices = choices.map((choice) => ({ ...choice, disabled: false }))
  }
  return isKubernetes(deployment) ? choices : choices.concat('none')
}

function getDownloadableChoices(versions) {
  return Object.entries(versions).map(([key, binaryInfo]) => ({
    name: binaryInfo.description,
    value: key,
    disabled: disableIfWrongJavaVersion(binaryInfo),
  }))
}

export function getTesseraOnPath() {
  const pathChoices = []
  const tesseraJarEnv = process.env.TESSERA_JAR
  if (tesseraJarEnv) {
    pathChoices.push({
      name: `Tessera at $TESSERA_JAR (${tesseraJarEnv})`,
      value: 'PATH',
    })
  }
  return pathChoices
}

export function pathToQuorumBinary(quorumVersion) {
  if (quorumVersion === 'PATH') {
    return 'geth'
  }
  const binary = BINARIES.quorum[quorumVersion]
  return joinPath(wizardHomeDir(), 'bin', 'quorum', quorumVersion, binary.name)
}

export function pathToTesseraJar(transactionManager) {
  if (transactionManager === 'PATH') {
    return '$TESSERA_JAR'
  }
  const binary = BINARIES.tessera[transactionManager]
  return joinPath(wizardHomeDir(), 'bin', 'tessera', transactionManager, binary.name)
}

export function pathToCakeshop(version) {
  const binary = BINARIES.cakeshop[version]
  return joinPath(wizardHomeDir(), 'bin', 'cakeshop', version, binary.name)
}

export function pathToIstanbulTools() {
  const binary = BINARIES.istanbul[LATEST_ISTANBUL_TOOLS]
  return joinPath(wizardHomeDir(), 'bin', 'istanbul', LATEST_ISTANBUL_TOOLS, binary.name)
}

export function pathToBootnode() {
  const binary = BINARIES.bootnode[LATEST_BOOTNODE]
  return joinPath(wizardHomeDir(), 'bin', 'bootnode', LATEST_BOOTNODE, binary.name)
}
