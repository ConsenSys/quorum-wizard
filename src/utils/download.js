import { basename, join } from 'path'
import { createFolder, cwd, exists } from './fileUtils'
import axios from 'axios'
import { createGunzip } from 'zlib'
import { extract } from 'tar-fs'
import { createWriteStream } from 'fs'
import { BINARIES, getPlatformSpecificUrl } from './binaryHelper'

export async function downloadIfMissing (name, version) {
  if (BINARIES[name] === undefined || BINARIES[name][version] === undefined) {
    throw new Error(`Could not find binary info entry for ${name} ${version}`)
  }
  const binaryInfo = BINARIES[name][version]
  const binDir = join(cwd(), 'bin', name, version)
  const binaryFileLocation = join(binDir, binaryInfo.name)
  if (!exists(binDir)) {
    createFolder(binDir, true)
    const url = getPlatformSpecificUrl(binaryInfo)

    console.log(`Downloading ${name} ${version} from ${url}...`)
    const response = await axios({
      url: url,
      method: 'GET',
      responseType: 'stream',
    })

    console.log(`Saving to ${binaryFileLocation}`)
    if (binaryInfo.type === 'tar.gz') {
      const extractorStream = response.data.pipe(createGunzip())
        .pipe(extract(binDir, {
          map: function (header) {
            const filename = basename(header.name)
            if (binaryInfo.files.includes(filename)) {
              // don't include folders when extracting files we want
              header.name = filename
            }
            return header
          },
          ignore: function (name) {
            return !binaryInfo.files.includes(basename(name))
          },
        }))
      return new Promise((resolve, reject) => {
        extractorStream.on('finish', () => {
          console.log('Done')
          resolve()
        })
        extractorStream.on('error', reject)
      })
    } else {
      const writer = createWriteStream(binaryFileLocation, { mode: 0o755 })
      response.data.pipe(writer)
      return new Promise((resolve, reject) => {
        writer.on('finish', () => {
          console.log('Done')
          resolve()
        })
        writer.on('error', reject)
      })
    }
  } else {
    console.log('Using cached binary at:', binaryFileLocation)
  }
}
