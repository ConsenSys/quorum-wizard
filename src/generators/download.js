import { basename } from 'path'
import axios from 'axios'
import { createGunzip } from 'zlib'
import { extract } from 'tar-fs'
import { createWriteStream } from 'fs'
import {
  createFolder,
  exists,
  wizardHomeDir,
} from '../utils/fileUtils'
import { info } from '../utils/log'
import { joinPath } from '../utils/pathUtils'

export async function getVersions(name) {
  const url = `https://api.bintray.com/packages/quorumengineering/${name}`
  const response = await axios({
    url,
    method: 'GET',
  })
  return response.data.versions
}

// eslint-disable-next-line import/prefer-default-export
export async function downloadIfMissing(name, version) {
  let binaryInfo
  switch (name) {
    case 'quorum':
      binaryInfo = createQuorumBinaryInfo(version)
      break
    case 'istanbul':
      binaryInfo = createIstanbulBinaryInfo(version)
      break
    case 'bootnode':
      binaryInfo = createBootnodeBinaryInfo(version)
      break
    default:
      binaryInfo = BINARIES[name][version]
  }
  const binDir = joinPath(wizardHomeDir(), 'bin', name, version)
  const binaryFileLocation = joinPath(binDir, binaryInfo.name)
  if (!exists(binaryFileLocation)) {
    createFolder(binDir, true)
    const url = getPlatformSpecificUrl(binaryInfo)

    info(`Downloading ${name} ${version} from ${url}...`)
    const response = await axios({
      url,
      method: 'GET',
      responseType: 'stream',
    })

    info(`Unpacking to ${binaryFileLocation}`)
    if (binaryInfo.type === 'tar.gz') {
      const extractorStream = response.data.pipe(createGunzip())
        .pipe(extract(binDir, {
          map: (header) => {
            const filename = basename(header.name)
            if (binaryInfo.files.includes(filename)) {
              // don't include folders when extracting files we want
              header.name = filename // eslint-disable-line no-param-reassign
            }
            return header
          },
          ignore: (pathName) => !binaryInfo.files.includes(basename(pathName)),
        }))
      return new Promise((resolve, reject) => {
        extractorStream.on('finish', () => {
          info(`Saved to ${binaryFileLocation}`)
          resolve()
        })
        extractorStream.on('error', reject)
      })
    }
    const writer = createWriteStream(binaryFileLocation, { mode: 0o755 })
    response.data.pipe(writer)
    return new Promise((resolve, reject) => {
      writer.on('finish', () => {
        info(`Saved to ${binaryFileLocation}`)
        resolve()
      })
      writer.on('error', reject)
    })
  }
  info(`Using cached ${name} at: ${binaryFileLocation}`)
  return binaryFileLocation
}

export function getPlatformSpecificUrl({ url }) {
  if (typeof url === 'string') {
    return url
  }
  const platformUrl = url[process.platform]
  if (platformUrl === undefined) {
    throw new Error(
      `Sorry, your platform (${process.platform}) is not supported.`,
    )
  }
  return platformUrl
}

export function createQuorumBinaryInfo(version) {
  return {
    name: 'geth',
    description: `Quorum ${version}`,
    url: {
      darwin: `https://bintray.com/quorumengineering/quorum/download_file?file_path=${version}/geth_${version}_darwin_amd64.tar.gz`,
      linux: `https://bintray.com/quorumengineering/quorum/download_file?file_path=${version}/geth_${version}_linux_amd64.tar.gz`,
    },
    type: 'tar.gz',
    files: [
      'geth',
    ],
  }
}

export function createIstanbulBinaryInfo(version) {
  return {
    name: 'istanbul',
    url: {
      darwin: `https://bintray.com/api/ui/download/quorumengineering/istanbul-tools/istanbul-tools_${version}_darwin_amd64.tar.gz`,
      linux: `https://bintray.com/api/ui/download/quorumengineering/istanbul-tools/istanbul-tools_${version}_linux_amd64.tar.gz`,
    },
    type: 'tar.gz',
    files: [
      'istanbul',
    ],
  }
}

export function createBootnodeBinaryInfo(version) {
  return {
    name: 'bootnode',
    url: {
      darwin: `https://bintray.com/api/ui/download/quorumengineering/geth-bootnode/bootnode_${version}_darwin_amd64.tar.gz`,
      linux: `https://bintray.com/api/ui/download/quorumengineering/geth-bootnode/bootnode_${version}_linux_amd64.tar.gz`,
    },
    type: 'tar.gz',
    files: [
      'bootnode',
    ],
  }
}

export const BINARIES = {
  tessera: {
    '0.10.4': {
      name: 'tessera-app.jar',
      description: 'Tessera 0.10.4',
      url: 'https://oss.sonatype.org/service/local/repositories/releases/content/com/jpmorgan/quorum/tessera-app/0.10.4/tessera-app-0.10.4-app.jar',
      type: 'jar',
    },
    '0.10.2': {
      name: 'tessera-app.jar',
      description: 'Tessera 0.10.2',
      url: 'https://oss.sonatype.org/service/local/repositories/releases/content/com/jpmorgan/quorum/tessera-app/0.10.2/tessera-app-0.10.2-app.jar',
      type: 'jar8',
    },
  },

  cakeshop: {
    '0.11.0': {
      name: 'cakeshop.war',
      description: 'Cakeshop 0.11.0',
      url: 'https://github.com/jpmorganchase/cakeshop/releases/download/v0.11.0/cakeshop-0.11.0.war',
      type: 'jar',
    },
    '0.11.0-J8': {
      name: 'cakeshop.war',
      description: 'Cakeshop 0.11.0-J8',
      url: 'https://github.com/jpmorganchase/cakeshop/releases/download/v0.11.0/cakeshop-0.11.0-J8.war',
      type: 'jar8',
    },
  },
}
