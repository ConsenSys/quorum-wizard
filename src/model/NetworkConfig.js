import { isJava11Plus } from '../utils/execUtils'
import {
  LATEST_CAKESHOP,
  LATEST_CAKESHOP_J8,
  LATEST_QUORUM,
  LATEST_TESSERA,
  LATEST_TESSERA_J8,
} from '../generators/download'
import {
  getDockerSubnet,
  cidrhost,
} from '../utils/subnetUtils'
import { getOutputPath } from '../utils/fileUtils'

export function createConfigFromAnswers(answers) {
  const {
    name,
    networkPath = getOutputPath(),
    numberNodes = 3,
    consensus = 'raft',
    quorumVersion = LATEST_QUORUM,
    transactionManager = isJava11Plus() ? LATEST_TESSERA : LATEST_TESSERA_J8,
    deployment = 'bash',
    cakeshop = isJava11Plus() ? LATEST_CAKESHOP : LATEST_CAKESHOP_J8,
    generateKeys = false,
    networkId = '10',
    genesisLocation = 'none',
    customizePorts = false,
    nodes = [],
    cakeshopPort = '8999',
    remoteDebug = false,
    containerPorts = undefined,
  } = answers
  const networkFolder = name
    || defaultNetworkName(numberNodes, consensus, transactionManager, deployment)
  const dockerSubnet = (isDocker(deployment) && containerPorts !== undefined) ? containerPorts.dockerSubnet : ''
  return {
    network: {
      name: networkFolder,
      networkPath,
      verbosity: 5,
      consensus,
      quorumVersion,
      transactionManager,
      permissioned: true,
      genesisFile: genesisLocation,
      generateKeys,
      configDir: `network/${networkFolder}/resources`,
      deployment,
      cakeshop,
      networkId,
      customizePorts,
      cakeshopPort,
      remoteDebug,
    },
    nodes: (customizePorts && nodes.length > 0) ? nodes : generateNodeConfigs(
      numberNodes,
      transactionManager,
      deployment,
      cakeshop,
      dockerSubnet,
    ),
    containerPorts,
  }
}

export function defaultNetworkName(numberNodes, consensus, transactionManager, deployment) {
  const transactionManagerName = !isTessera(transactionManager)
    ? ''
    : 'tessera-'
  return `${numberNodes}-nodes-${consensus}-${transactionManagerName}${deployment}`
}

export function generateNodeConfigs(
  numberNodes,
  transactionManager,
  deployment,
  cakeshop,
  dockerSubnet,
) {
  const devP2pPort = 21000
  const rpcPort = 22000
  const wsPort = 23000
  const graphQlPort = 24000
  const raftPort = 50401
  const thirdPartyPort = 9081
  const p2pPort = 9001
  const nodes = []

  for (let i = 0; i < parseInt(numberNodes, 10); i += 1) {
    const node = {
      quorum: {
        ip: isDocker(deployment) ? cidrhost(dockerSubnet, i + 1 + 10) : '127.0.0.1',
        devP2pPort: devP2pPort + i,
        rpcPort: rpcPort + i,
        wsPort: wsPort + i,
        raftPort: raftPort + i,
        graphQlPort: graphQlPort + i,
      },
    }
    if (isTessera(transactionManager)) {
      node.tm = {
        ip: isDocker(deployment) ? cidrhost(dockerSubnet, i + 1 + 100) : '127.0.0.1',
        thirdPartyPort: thirdPartyPort + i,
        p2pPort: p2pPort + i,
      }
    }
    nodes.push(node)
  }
  return nodes
}

export function getContainerPorts(deployment) {
  const dockerSubnet = isDocker(deployment) ? getDockerSubnet() : ''
  return {
    dockerSubnet,
    quorum: {
      rpcPort: 8545,
      p2pPort: 21000,
      raftPort: 50400,
      wsPort: 8645,
      graphQlPort: 8547,
    },
    tm: {
      p2pPort: 9000,
      thirdPartyPort: 9080,
    },
  }
}

export function isTessera(tessera) {
  return tessera !== 'none'
}

export function isDocker(deployment) {
  return deployment === 'docker-compose'
}

export function isBash(deployment) {
  return deployment === 'bash'
}

export function isKubernetes(deployment) {
  return deployment === 'kubernetes'
}

export function isIstanbul(consensus) {
  return consensus === 'istanbul'
}

export function isRaft(consensus) {
  return consensus === 'raft'
}

export function isCakeshop(cakeshop) {
  return cakeshop !== 'none'
}
