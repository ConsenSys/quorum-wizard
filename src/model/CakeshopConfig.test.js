import { createConfigFromAnswers } from './NetworkConfig'
import { generateCakeshopConfig } from './CakeshopConfig'
import { LATEST_CAKESHOP, LATEST_QUORUM, LATEST_TESSERA } from '../generators/download'
import { cidrhost } from '../utils/subnetUtils'

jest.mock('../utils/subnetUtils')
cidrhost.mockReturnValue('172.16.239.11')

const baseNetwork = {
  numberNodes: '3',
  consensus: 'raft',
  quorumVersion: LATEST_QUORUM,
  transactionManager: LATEST_TESSERA,
  deployment: 'bash',
  cakeshop: LATEST_CAKESHOP,
}

test('creates 3nodes raft dockerFile tessera cakeshop', () => {
  const config = createConfigFromAnswers({
    ...baseNetwork,
    deployment: 'docker-compose',
  })
  const cakeshop = generateCakeshopConfig(config)
  expect(cakeshop).toMatchSnapshot()
})

test('creates 3nodes istanbul bash tessera cakeshop', () => {
  const config = createConfigFromAnswers({
    ...baseNetwork,
    consensus: 'istanbul',
  })
  const cakeshop = generateCakeshopConfig(config)
  expect(cakeshop).toMatchSnapshot()
})

test('creates 3nodes raft dockerFile no tessera cakeshop', () => {
  const config = createConfigFromAnswers({
    ...baseNetwork,
    transactionManager: 'none',
    deployment: 'docker-compose',
  })
  const cakeshop = generateCakeshopConfig(config)
  expect(cakeshop).toMatchSnapshot()
})

test('creates 3nodes istanbul bash no tessera cakeshop', () => {
  const config = createConfigFromAnswers({
    ...baseNetwork,
    consensus: 'istanbul',
    transactionManager: 'none',
  })
  const cakeshop = generateCakeshopConfig(config)
  expect(cakeshop).toMatchSnapshot()
})
