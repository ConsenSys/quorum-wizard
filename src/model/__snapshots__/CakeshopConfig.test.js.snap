// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates 3nodes istanbul bash no tessera cakeshop 1`] = `
Array [
  Object {
    "name": "node1",
    "rpcUrl": "http://localhost:22000",
    "transactionManagerUrl": "",
  },
  Object {
    "name": "node2",
    "rpcUrl": "http://localhost:22001",
    "transactionManagerUrl": "",
  },
  Object {
    "name": "node3",
    "rpcUrl": "http://localhost:22002",
    "transactionManagerUrl": "",
  },
]
`;

exports[`creates 3nodes istanbul bash tessera cakeshop 1`] = `
Array [
  Object {
    "name": "node1",
    "rpcUrl": "http://localhost:22000",
    "transactionManagerUrl": "http://localhost:9081",
  },
  Object {
    "name": "node2",
    "rpcUrl": "http://localhost:22001",
    "transactionManagerUrl": "http://localhost:9082",
  },
  Object {
    "name": "node3",
    "rpcUrl": "http://localhost:22002",
    "transactionManagerUrl": "http://localhost:9083",
  },
]
`;

exports[`creates 3nodes raft dockerFile no tessera cakeshop 1`] = `
Array [
  Object {
    "name": "node1",
    "rpcUrl": "http://node1:8545",
    "transactionManagerUrl": "",
  },
  Object {
    "name": "node2",
    "rpcUrl": "http://node2:8545",
    "transactionManagerUrl": "",
  },
  Object {
    "name": "node3",
    "rpcUrl": "http://node3:8545",
    "transactionManagerUrl": "",
  },
]
`;

exports[`creates 3nodes raft dockerFile tessera cakeshop 1`] = `
Array [
  Object {
    "name": "node1",
    "rpcUrl": "http://node1:8545",
    "transactionManagerUrl": "http://txmanager1:9080",
  },
  Object {
    "name": "node2",
    "rpcUrl": "http://node2:8545",
    "transactionManagerUrl": "http://txmanager2:9080",
  },
  Object {
    "name": "node3",
    "rpcUrl": "http://node3:8545",
    "transactionManagerUrl": "http://txmanager3:9080",
  },
]
`;
